R/utils.R

```diff
@@ -11,132 +9,113 @@
 # `write` operation) to be undone, AND for any future `writeWS` calls to be
 # ignored. A call to `write` will be respected, and will restore normal
 # behavior.
-WSTextWriter <- R6Class("WSTextWriter",
-  private = list(
-    writer = "TextWriter",
-    suppressing = logical(1)
-  ),
-  public = list(
-    initialize = function() {
-      private$writer <- TextWriter$new()
-      private$suppressing <- FALSE
-    },
-    close = function() {
-      private$writer$close()
-    },
-    write = function(text) {
-      private$writer$write(text)
+#
+# Text is automatically converted to UTF-8 before being written.
+#' @param bufferSize The initial size of the buffer in which writes are stored.
+#'   The buffer will be periodically cleared, if possible, to cache the writes
+#'   as a string. If the buffer cannot be cleared (because of the need to be
+#'   able to backtrack to fulfill an `eatWS()` call), then the buffer size will
+#'   be doubled.
+#' @noRd
+#' @importFrom R6 R6Class
+WSTextWriter <- function(bufferSize=1024){
+  # A string into which we can collapse entries in the buffer if the buffer gets full.
+  accumulated <- ""
+
+  # The buffer into which we enter all the writes.
+  buffer <- character(bufferSize)
+
+  # The index storing the position in the buffer of the most recent write.
+  marked <- 0
+
+  # The index storing the position in the buffer of the most recent write or writeWS.
+  position <- 0
+
+  # TRUE if we're eating whitespace right now, in which case calls to eatWS are no-ops.
+  suppressing <- FALSE
+
+  # Logic to do the actual write
+  writeImpl <- function(text) {
+    # If an error is going to happen while evaluating `text`, let's make it
+    # happen before we mutate any of our internal state
+    force(text)
 
-      # Reset suppressing behavior
-      private$suppressing <- FALSE
+    # FIXME: test
+    # Are we at the end of our buffer?
+    if (position == length(buffer)) {
 
-      # Set a bookmark here, so that future calls to eatWS() don't affect this
-      # write
-      private$writer$savePosition()
+      # If our `marked` position has advanced well into the buffer,
+      # we can go ahead and collapse the string up to the marked position
+      # to free up room in the buffer.
+      if (marked >= length(buffer)/2) {
+        # Collapse the writes in the buffer up to the marked position
+        str <- paste(buffer[1:marked], collapse="")
```

Comment:
Better to use `seq_len(marked)` for safety. (I know that in this case, `marked` is definitely  >= 1, but it's still good practice).